text/0076-entity-slice-validation.md

```diff
@@ -0,0 +1,443 @@
+permit (
+  principal in Team::"Admin",
+  action,
+  resource
+);
+```
+
+We suppose (per the [schema](https://github.com/cedar-policy/cedar-examples/blob/release/3.2.x/tinytodo/tinytodo.cedarschema), not shown) that `principal` always has entity type `User`, and `resource` always has entity type `List`. These policies are valid at level 1, meaning that variables `principal`, `action`, `resource` and `context` are only *directly* dereferenced, i.e., any entities they reference are not themselves dereferenced.
+
+For example, in policy 1, the expression `resource.owner == principal` directly dereferences `resource` by accessing the contents of its `owner` attribute. In policy 2, expression `principal in resource.readers` directly dereferences `resource` to retrieve the contents of its `readers` attribute, and also directly dereferences `principal` to retrieve its ancestors, to see whether one might be equal to the contents of `resource.readers`. Policy 4 also directly dereferences `principal` to access its ancestors, to see if one is `Team::"admin"`.
+
+Because these policies are valid at level 1, applications submitting authorization requests only need to provide entity data for entities provided as variables, and nothing else. For example, say we want to evaluate the request (`User::"Aaron"`, `Action::"GetList"`, `List::"Objectives"`, `{}`). Then we would only need to provide entity data (direct attributes and ancestors) for `User::"Aaron"`, `Action::"GetList"`, and `List::"Objectives"`. There is no need to provide data for entities referenced from these entities. For example, suppose the `readers` attribute of `List::"Objectives"` is `Team::"interns"`; there is no need to provide data for `Team::"interns"` because we know it will never, itself, be dereferenced when evaluating these policies. There is also no need to provide entity data for other entities directly mentioned in policies, e.g., `Team::"Admin"`.
+
+Now suppose we wanted to extend our TinyTodo policies with the following:
+
+```
+// Policy 6: No access if not high rank and at location DEF, 
+// or at resource's owner's location
+forbid(
+    principal,
+    action,
+    resource is List
+) unless {
+   principal.joblevel > 6 && principal.location like "DEF*" ||
+   principal.location == resource.owner.location
+};
+```
+
+This policy does *not* validate at level 1. That is because the expression `resource.owner.location` is only valid at level 2: It requires dereferencing `resource` to retrieve the entity in the `owner` attribute, and then requires dereferencing that entity to obtain the contents of its `location` attribute. As a result, if we only provided the `principal` and `resource` entity data, as described above, evaluating policy 6 could produce an evaluation error since data may not be provided for `resource.owner`.
+
+All of the Cedar example policy sets validate with level-based validation, either at level 1 or 2; see the Appendix for details.
+
+## Motivation
```

Comment:
Also, I'm curious if there are real-world examples which motivated this proposal. If so, I think it would be useful to include them.

Reply:
I agree that #74 is likely to be more suitable for most consumers of Cedar. This RFC gives a less granular tool for describing the entity data that needs to be included in the slice. Sometimes, being more granular is better.

The usecase I find most compelling for this RFC is one where one party writes the entity slicing algorithm and another writes the policies (e.g., you offer a service where end-users can author Cedar policies). With #74, the entity slicing potentially needs to be changed whenever policies change (because we were passing in only the entities needed). With this RFC, we pass in some entity data that may be unused, but this means policies can change freely as long as they continue to only use the entity data at level `n`.

Reply:
"End-users writing policies" is (in vague terms) the real-world example that motivated this.